### ISSUE TITLE
[Timestampable] Existing null values from v1 throw exceptions

### ISSUE THREAD

**Issue by spackmat** [KnpLabs/DoctrineBehaviors#496]
I migrated a project from v1.6 to v2.0.1 an get a lot of Exeptions from the Timestampable Behavior:

> Return value of App\Entity\MyEntity::getUpdatedAt() must implement interface DateTimeInterface, null returned

Indeed, `getUpdatedAt()` and the other getters have a signature that allows only `DateTimeInterface` as a return value, but my existing data has always `null` for all entities that were never updated.

I'm not sure about side effects and why the signature was changed that way, but I think `?DateTimeInterface` as a return type hint would solve that.

**Comment by TomasVotruba** [comment]
>  but I think ?DateTimeInterface as a return type hint would solve that.

:+1:

**Comment by spackmat** [comment]
When I see such strict signatures, I always assume that it is was introduced on purpose, so in that case it maybe would be better to just return the value of `createdAt`, when `updatedAt` is `null`?

**Comment by TomasVotruba** [comment]
It was on purpose within scope of this library's tsets, but it didn't cover all the possible cases. 

> Better to just return the value of createdAt, when updatedAt is null?

That could make sense.

**Comment by spackmat** [comment]
That would be a BC break as I used updatedAt is null to mark entities as never updated for the users. I could change that, but it's not clear to me, why updatedAt should not be null. You say because the tests say so. But why did you change that at all from v1?

I made the return type hint nullable in the interface and the implementation and that works fine as far as I could see. That solution wouldn't be a BC break. Should I send a PR for that?

**Comment by TomasVotruba** [comment]
Yes, I agreed to PR in previous comment:
https://github.com/KnpLabs/DoctrineBehaviors/issues/496#issuecomment-572951587

**Comment by spackmat** [comment]
Here it is. But to be honest, I'm not sure, if the decision to always populate `updatedAt` is a good one.

The upside ist, that one only has to to look into that property and always get valid data.

The downsides are that it must have an inconsistent method signature to not break legacy data, stores redundant data for not-updated entities and it is a bit more complicated to check, if an entity was ever updated (besides that populating `updatedAt` is simply not correct, when an entity was in fact never updated).

So as an alternative to my PR I see two possible solutions: Update the legacy data to not have null values in the `updatedAt` column or reverting the decision and behave like v1 again, not touching `updatedAt` on persist.

**Comment by TomasVotruba** [comment]
> ...the decision to always populate updatedAt is a good one.

That's convention here, changing it would break lot of use cases.
Saying that, I think it's a good one. If you need conditional update of updatedAt, that's an edge case and it's better to use own method with life cycle callbacks that you have under control.

This extension should help with simple repeated use cases, to delegate complexity out of the application and descrease error rate.

**Comment by spackmat** [comment]
> That's convention here, changing it would break lot of use cases.

It wasn't in v1, so introducing it as a convention for v2 breaks BC. You think this is worth it and the new behavior is better, you are the new maintainer and have a better overview, that't acceptable..

I'll write a migration and update all my legacy data to cope with that (or invest the time to get the bundle out of my project as I only use blameable and timestampable that are quite simple to implement).

Perhaps that need for data migration needs to be documented in the migration from v1 to v2 docs at it hits everyone.

But I cannot accept that leaving an `updatedAt` property at `null` unless the entity gets updated the first time is considered an "edge case". Or our understanding of an edge case differs quite a lot.

**Comment by spackmat** [comment]
Just for the record: While I implemented the simple Blameable/Timestampable in my own code (really quite simple, no bundle needed for that), I came across several existing datasets that also had the `createdAt`-property with a value of `null`. This was allowed before in v1 and all datasets before implementation of the Timestampable are affected. To migrate that data to v2, a suitable default date must be defined. That means, I cannot write and distribute a migration script, because someone with overview over the business data has to decide what timestamp is suitable for those existing datasets.

I strongly suggest to point that out loud in the migration section of the README. To migrate existing data to v2 of the DoctrineBehaviors bundle, it is necessary to:

1. Find all datasets with `created_at IS NULL` and fill this with a suitable date. This is a decision about existing business data, a dev maybe doesn't want to make alone and for some projects an automatic migration script cannot be provided.
2. Find all datasets with `updated_at IS NULL` and copy the value from `created_at` over. This can be automated with a migration (after 1. is resolved).

The Blameable properties are not that problematic as they are still nullable in v2 (mostly because they can be a string or a relation to an external entity which prevents generic return type hints). This alone can be a good reason to implement the behavior for the individual usecase.
